Autopilot/PathManager/Src/waypointManager.cpp

```diff
@@ -0,0 +1,952 @@
+        float dotProduct = waypointDirection[0] * (position[0] - halfPlane[0]) + waypointDirection[1] * (position[1] - halfPlane[1]) + waypointDirection[2] * (position[2] - halfPlane[2]);
+        
+        if (dotProduct > 0){
+            orbitPathStatus = ORBIT_FOLLOW;
+            if (targetWaypoint->waypointType == HOLD_WAYPOINT) {
+                inHold = true;
+                turnDirection = 1; // Automatically turn CCW
+                turnRadius = targetWaypoint->turnRadius;
+                turnDesiredAltitude = targetWaypoint->altitude;
+                turnCenter[0] = targetWaypoint->longitude;
+                turnCenter[1] = targetWaypoint->latitude;
+                turnCenter[2] = turnDesiredAltitude;
+                /*
+                    Increment the currentIndex so the plane is not perpetually stuck in a holding pattern
+                */
+                currentIndex++; 
+            }
+        }
+
+        // std::cout << "Here2 --> " << position[0] << " " << position[1] << " " << position[2] << std::endl;
+        // std::cout << "Here3 --> " << waypointDirection[0] << " " << waypointDirection[1] << " " << waypointDirection[2] << std::endl;
+        // std::cout << "Here4 --> " << targetCoordinates[0] << " " << targetCoordinates[1] << " " << targetCoordinates[2] << std::endl;
+
+        follow_straight_path(waypointDirection, targetCoordinates, position, heading);
+    } else {
+        // Determines turn direction (CCW returns 2; CW returns 1)
+        turnDirection = waypointDirection[0] * nextWaypointDirection[1] - waypointDirection[1] * nextWaypointDirection[0]>0?1:-1;
+        
+        // Since the Earth is not flat *waits for the uproar to die down* we need to do some fancy geometry. Introducing!!!!!!!!!! EUCLIDIAN GEOMETRY! (translation: I have no idea what this line does but it should work)
+        float euclideanWaypointDirection = sqrt(pow(nextWaypointDirection[0] - waypointDirection[0],2) + pow(nextWaypointDirection[1] - waypointDirection[1],2) + pow(nextWaypointDirection[2] - waypointDirection[2],2)) * ((nextWaypointDirection[0] - waypointDirection[0]) < 0?-1:1) * ((nextWaypointDirection[1] - waypointDirection[1]) < 0?-1:1) * ((nextWaypointDirection[2] - waypointDirection[2]) < 0?-1:1);
+
+        // Determines coordinates of the turn center
+        turnCenter[0] = targetCoordinates[0] + (tangentFactor * (nextWaypointDirection[0] - waypointDirection[0])/euclideanWaypointDirection);
+        turnCenter[1] = targetCoordinates[1] + (tangentFactor * (nextWaypointDirection[1] - waypointDirection[1])/euclideanWaypointDirection);
+        turnCenter[2] = targetCoordinates[2] + (tangentFactor * (nextWaypointDirection[2] - waypointDirection[2])/euclideanWaypointDirection);
+
+        float dotProduct = nextWaypointDirection[0] * (position[0] - halfPlane[0]) + nextWaypointDirection[1] * (position[1] - halfPlane[1]) + nextWaypointDirection[2] * (position[2] - halfPlane[2]);
+        
+        if (dotProduct > 0){
```

Comment:
Come to think of it, this might be susceptible to noise. Our sensors aren't perfect, so we might find that the dot product grows larger than 0prematurely because of that jitter. Though this is something to think about in the future rather than in this PR.